File: backend/davon-library-backend/src/main/java/org/acme/service/ReservationService.java
package org.acme.service;

import org.acme.model.Book;
import org.acme.model.BookStatus;
import org.acme.model.Reservation;
import org.acme.model.ReservationStatus;
import org.acme.model.Loan;
import org.acme.repository.BookRepository;
import org.acme.repository.ReservationRepository;
import org.acme.repository.UserRepository;
import org.acme.repository.LoanRepository;

import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import jakarta.transaction.Transactional;
import jakarta.ws.rs.BadRequestException;
import jakarta.ws.rs.ForbiddenException;
import jakarta.ws.rs.NotFoundException;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;
import java.util.Objects;
import org.acme.model.User;

/**
 * Service for managing book reservations.
 * This service handles creation, cancellation, and retrieval of book
 * reservations.
 */
@ApplicationScoped
public class ReservationService {

    private final ReservationRepository reservationRepository;
    private final BookRepository bookRepository;
    private final UserRepository userRepository;
    private final LoanRepository loanRepository;
    private final NotificationService notificationService;

    private static final int RESERVATION_EXPIRATION_DAYS = 7;
    private static final int MAX_ACTIVE_RESERVATIONS_PER_USER = 5;

    @Inject
    public ReservationService(ReservationRepository reservationRepository, BookRepository bookRepository,
            UserRepository userRepository, LoanRepository loanRepository, NotificationService notificationService) {
        this.reservationRepository = reservationRepository;
        this.bookRepository = bookRepository;
        this.userRepository = userRepository;
        this.loanRepository = loanRepository;
        this.notificationService = notificationService;
    }

    /**
     * Creates a new reservation for a book by a user.
     *
     * @param bookId The ID of the book to reserve.
     * @param userId The ID of the user making the reservation.
     * @return The newly created Reservation object.
     * @throws NotFoundException     if the book is not found.
     * @throws IllegalStateException if the book is not available for reservation.
     * @throws BadRequestException   if the user has reached the maximum number of
     *                               active reservations.
     */
    @Transactional
    public Reservation createReservation(Long bookId, Long userId) {
        Objects.requireNonNull(bookId, "Book ID cannot be null");
        Objects.requireNonNull(userId, "User ID cannot be null");

        Book book = bookRepository.findByIdOptional(bookId)
                .orElseThrow(() -> new NotFoundException("Book not found with ID: " + bookId));

        if (book.getStatus() != BookStatus.AVAILABLE && book.getStatus() != BookStatus.BORROWED) {
            throw new IllegalStateException("Book is not available for reservation.");
        }

        if (!canUserReserveBook(userId, bookId)) {
            throw new BadRequestException(
                    "User has reached the maximum number of active reservations or has already reserved this book.");
        }

        Reservation reservation = new Reservation();

        // Set the object references instead of IDs
        reservation.setBook(book);
        User user = userRepository.findByIdOptional(userId)
                .orElseThrow(() -> new NotFoundException("User not found with ID: " + userId));
        reservation.setUser(user);

        reservation.setReservationDate(LocalDate.now());
        reservation.setExpirationDate(LocalDate.now().plusDays(RESERVATION_EXPIRATION_DAYS));
        reservation.setStatus(ReservationStatus.ACTIVE);

        // Calculate queue position
        int queuePosition = getNextQueuePosition(bookId);
        reservation.setQueuePosition(queuePosition);

        reservation.setCreatedAt(LocalDateTime.now());
        reservation.setUpdatedAt(LocalDateTime.now());

        // Only change status to RESERVED if book was AVAILABLE
        // BORROWED books should remain BORROWED even when reserved
        if (book.getStatus() == BookStatus.AVAILABLE) {
            book.setStatus(BookStatus.RESERVED);
            bookRepository.persist(book);
        }

        reservationRepository.persist(reservation);

        // Create notification for book reserved
        notificationService.createBookReservedNotification(userId, book.getTitle());

        return reservation;
    }

    /**
     * Cancels an existing reservation.
     *
     * @param reservationId The ID of the reservation to cancel.
     * @param userId        The ID of the user attempting to cancel the reservation.
     * @return The updated Reservation object with a CANCELLED status.
     * @throws NotFoundException  if the reservation is not found.
     * @throws ForbiddenException if the user does not own the reservation.
     */
    @Transactional
    public Reservation cancelReservation(Long reservationId, Long userId) {
        Objects.requireNonNull(reservationId, "Reservation ID cannot be null");
        Objects.requireNonNull(userId, "User ID cannot be null");

        Reservation reservation = reservationRepository.findByIdOptional(reservationId)
                .orElseThrow(() -> new NotFoundException("Reservation not found with ID: " + reservationId));

        if (!reservation.getUser().getId().equals(userId)) {
            throw new ForbiddenException("User does not have permission to cancel this reservation.");
        }

        reservation.setStatus(ReservationStatus.CANCELLED);
        reservation.setUpdatedAt(LocalDateTime.now());

        Book book = reservation.getBook();

        // Check if there are any active loans for this book
        List<Loan> activeLoans = loanRepository.list("book.id = ?1 and returnDate is null", book.getId());

        // Only set book to AVAILABLE if there are no active loans
        if (activeLoans.isEmpty()) {
            book.setStatus(BookStatus.AVAILABLE);
            bookRepository.persist(book);
        }
        // If there are active loans, the book should remain BORROWED (don't change
        // status)

        reservationRepository.persist(reservation);

        // Create notification for cancelled reservation
        notificationService.createReservationCancelledNotification(userId, book.getTitle());

        return reservation;
    }

    /**
     * Retrieves all active reservations for a specific user.
     *
     * @param userId The ID of the user.
     * @return A list of active reservations.
     */
    public List<Reservation> getUserActiveReservations(Long userId) {
        Objects.requireNonNull(userId, "User ID cannot be null");
        return reservationRepository.find("user.id = ?1 and status = ?2", userId, ReservationStatus.ACTIVE).list();
    }

    /**
     * Expires all active reservations that are past their expiration date.
     * This method is typically called by a scheduled job.
     */
    @Transactional
    public void expireOldReservations() {
        List<Reservation> reservationsToExpire = reservationRepository
                .find("status = ?1 and expirationDate < ?2", ReservationStatus.ACTIVE, LocalDate.now())
                .list();

        for (Reservation reservation : reservationsToExpire) {
            reservation.setStatus(ReservationStatus.EXPIRED);
            reservation.setUpdatedAt(LocalDateTime.now());

            Book book = reservation.getBook();

            // Check if there are any active loans for this book
            List<Loan> activeLoans = loanRepository.list("book.id = ?1 and returnDate is null", book.getId());

            // Only set book to AVAILABLE if there are no active loans
            if (activeLoans.isEmpty()) {
                book.setStatus(BookStatus.AVAILABLE);
                bookRepository.persist(book);
            }
            // If there are active loans, the book should remain BORROWED (don't change
            // status)

            reservationRepository.persist(reservation);
        }
    }

    /**
     * Retrieves the reservation queue for a specific book, sorted by reservation
     * date.
     *
     * @param bookId The ID of the book.
     * @return A sorted list of reservations for the book.
     */
    public List<Reservation> getBookReservationQueue(Long bookId) {
        Objects.requireNonNull(bookId, "Book ID cannot be null");
        return reservationRepository.find("book.id", bookId).list();
    }

    /**
     * Checks if a user is allowed to reserve a specific book.
     *
     * @param userId The ID of the user.
     * @param bookId The ID of the book.
     * @return true if the user can reserve the book, false otherwise.
     */
    private boolean canUserReserveBook(Long userId, Long bookId) {
        long activeReservations = reservationRepository
                .count("user.id = ?1 and status = ?2", userId, ReservationStatus.ACTIVE);

        if (activeReservations >= MAX_ACTIVE_RESERVATIONS_PER_USER) {
            return false;
        }

        long existingReservationForBook = reservationRepository
                .count("user.id = ?1 and book.id = ?2 and status = ?3", userId, bookId, ReservationStatus.ACTIVE);

        return existingReservationForBook == 0;
    }

    /**
     * Calculate the next queue position for a book reservation.
     */
    private int getNextQueuePosition(Long bookId) {
        List<Reservation> activeReservations = reservationRepository.list(
                "book.id = ?1 AND status = ?2 ORDER BY createdAt ASC",
                bookId, ReservationStatus.ACTIVE);
        return activeReservations.size() + 1;
    }

    /**
     * Fix inconsistent reservation states - marks reservations as FULFILLED
     * when user has an active loan for the same book.
     * This is a utility method to fix existing data.
     */
    @Transactional
    public int fixInconsistentReservations() {
        // Find reservations that are ACTIVE but user has an active loan for the same
        // book
        List<Reservation> inconsistentReservations = reservationRepository.find(
                "status = ?1 AND EXISTS (SELECT 1 FROM Loan l WHERE l.book.id = book.id AND l.user.id = user.id AND l.returnDate IS NULL)",
                ReservationStatus.ACTIVE).list();

        int updatedCount = 0;
        for (Reservation reservation : inconsistentReservations) {
            reservation.setStatus(ReservationStatus.FULFILLED);
            reservation.setUpdatedAt(LocalDateTime.now());
            reservationRepository.persist(reservation);
            updatedCount++;
        }

        return updatedCount;
    }
}

File: frontend/davon-library-webui/src/lib/api/services/reservation.service.ts
import { httpClient } from '../client';
import { API_CONFIG } from '../config';
import { Reservation, User, Book } from '../types';

export interface ReservationCreateRequest {
  bookId: number;
  userId: number;
}

export class ReservationService {
  /**
   * Create a new reservation
   */
  async createReservation(bookId: number, userId: number): Promise<Reservation> {
    try {
      return await httpClient.post<Reservation>(
        `/api/reservations/book/${bookId}/user/${userId}`,
        {}
      );
    } catch (error) {
      console.error(`Failed to create reservation for book ${bookId}:`, error);
      throw error;
    }
  }

  /**
   * Cancel a reservation
   */
  async cancelReservation(reservationId: number, userId: number): Promise<Reservation> {
    try {
      return await httpClient.delete<Reservation>(
        `/api/reservations/${reservationId}/user/${userId}`
      );
    } catch (error) {
      console.error(`Failed to cancel reservation ${reservationId}:`, error);
      throw error;
    }
  }

  /**
   * Get user's active reservations
   */
  async getUserReservations(userId: number): Promise<Reservation[]> {
    try {
      return await httpClient.get<Reservation[]>(
        API_CONFIG.ENDPOINTS.RESERVATIONS.BY_USER(userId)
      );
    } catch (error) {
      console.error(`Failed to fetch reservations for user ${userId}:`, error);
      throw error;
    }
  }

  /**
   * Get reservation queue for a specific book
   */
  async getBookReservationQueue(bookId: number): Promise<Reservation[]> {
    try {
      return await httpClient.get<Reservation[]>(
        `/api/reservations/book/${bookId}/queue`
      );
    } catch (error) {
      console.error(`Failed to fetch reservation queue for book ${bookId}:`, error);
      throw error;
    }
  }

  /**
   * Get all reservations (admin only)
   */
  async getAllReservations(): Promise<Reservation[]> {
    try {
      return await httpClient.get<Reservation[]>(API_CONFIG.ENDPOINTS.RESERVATIONS.BASE);
    } catch (error) {
      console.error('Failed to fetch all reservations:', error);
      throw error;
    }
  }

  /**
   * Check if user has already reserved a book
   */
  async hasUserReservedBook(userId: number, bookId: number): Promise<boolean> {
    try {
      const userReservations = await this.getUserReservations(userId);
      return userReservations.some(reservation => 
        reservation.book.id === bookId && 
        reservation.status === 'ACTIVE'
      );
    } catch (error) {
      console.error('Failed to check user reservation status:', error);
      return false;
    }
  }

  /**
   * Get user's position in reservation queue
   */
  async getUserPositionInQueue(userId: number, bookId: number): Promise<number> {
    try {
      const queue = await this.getBookReservationQueue(bookId);
      const position = queue.findIndex(reservation => reservation.user.id === userId);
      return position >= 0 ? position + 1 : 0; // Return 1-based position, 0 if not found
    } catch (error) {
      console.error('Failed to get user position in queue:', error);
      return 0;
    }
  }

  /**
   * Get reservation statistics
   */
  async getReservationStats(): Promise<{
    totalActive: number;
    totalExpired: number;
    totalFulfilled: number;
    totalCancelled: number;
  }> {
    try {
      const reservations = await this.getAllReservations();
      
      return {
        totalActive: reservations.filter(r => r.status === 'ACTIVE').length,
        totalExpired: reservations.filter(r => r.status === 'EXPIRED').length,
        totalFulfilled: reservations.filter(r => r.status === 'FULFILLED').length,
        totalCancelled: reservations.filter(r => r.status === 'CANCELLED').length,
      };
    } catch (error) {
      console.error('Failed to get reservation statistics:', error);
      return {
        totalActive: 0,
        totalExpired: 0,
        totalFulfilled: 0,
        totalCancelled: 0,
      };
    }
  }

  /**
   * Expire old reservations (admin function)
   */
  async expireOldReservations(): Promise<void> {
    try {
      await httpClient.post<void>('/api/reservations/expire-old', {});
    } catch (error) {
      console.error('Failed to expire old reservations:', error);
      throw error;
    }
  }

  /**
   * Check if a book can be reserved
   */
  async canReserveBook(bookId: number, userId: number): Promise<{
    canReserve: boolean;
    reason?: string;
  }> {
    try {
      // Check if user already has this book reserved
      const hasReserved = await this.hasUserReservedBook(userId, bookId);
      if (hasReserved) {
        return {
          canReserve: false,
          reason: 'You have already reserved this book'
        };
      }

      // Additional validation can be added here
      return { canReserve: true };
    } catch (error) {
      console.error('Failed to check if book can be reserved:', error);
      return {
        canReserve: false,
        reason: 'Unable to check reservation status'
      };
    }
  }
}

// Create and export singleton instance
export const reservationService = new ReservationService(); 